crates/runtime/src/vmcontext.rs

```diff
@@ -577,22 +618,37 @@ impl Default for VMSharedSignatureIndex {
     }
 }
 
-/// The VM caller-checked "anyfunc" record, for caller-side signature checking.
-/// It consists of the actual function pointer and a signature id to be checked
-/// by the caller.
+/// The VM caller-checked "funcref" record, for caller-side signature checking.
+///
+/// It consists of function pointer(s), a signature id to be checked by the
+/// caller, and the vmctx closure associated with this function.
 #[derive(Debug, Clone)]
 #[repr(C)]
 pub struct VMCallerCheckedFuncRef {
-    /// Function body.
-    pub func_ptr: NonNull<VMFunctionBody>,
+    /// Function pointer for this funcref if being called via the native calling
+    /// convention.
+    pub native_call: NonNull<VMNativeCallFunction>,
+
+    /// Function pointer for this funcref if being called via the "array"
+    /// calling convention that `Func::new` et al use.
+    pub array_call: VMArrayCallFunction,
+
+    /// Function pointer for this funcref if being called via the calling
+    /// convention we use when compiling Wasm.
+    ///
+    /// This can be null if no Wasm is loaded in the system or this `funcref`
+    /// hasn't been exposed to Wasm at all.
+    pub wasm_call: Option<NonNull<VMWasmCallFunction>>,
```

Comment:
The large doc block above about why this can be `None` I think might be good to insert here instead?